index.js

```diff
@@ -439,26 +472,29 @@ module.exports = function AsyncAppendOnlyLog(filename, opts) {
       return
     }
     onDrain(function startCompactAfterDrain() {
-      compaction = new Compaction(self, (err, sizeDiff) => {
-        compaction = null
-        if (err) return cb(err)
-        compactionProgress.set({ sizeDiff, percent: 1, done: true })
-        for (let i = 0, n = waitingCompaction.length; i < n; ++i) {
-          waitingCompaction[i]()
-        }
-        waitingCompaction.length = 0
-        cb()
-      })
-      compaction.progress((stats) => {
-        compactionProgress.set({ ...stats, done: false })
+      onDeletesFlushed(function startCompactAfterDeletes() {
+        compaction = new Compaction(self, (err, sizeDiff) => {
+          compaction = null
+          if (err) return cb(err)
+          compactionProgress.set({ sizeDiff, percent: 1, done: true })
+          for (const callback of waitingCompaction) callback()
+          waitingCompaction.length = 0
+          cb()
+        })
+        compaction.progress((stats) => {
+          compactionProgress.set({ ...stats, done: false })
+        })
```

Comment:
Probably hard to read this diff. It just adds on `onDeletesFlushed` after the onDrain, and then rewrites the for-loop with waitingCompaction, that's it.